### ISSUE TITLE
[Question] Is it safe to fit-transform once to get multiple embeddings?

### ISSUE THREAD

**Issue by jlopezpena** [lmcinnes/umap#226]
Quick question, a bit of the mathsy side.

Suppose I want to test different embeddings of my data, say in 2, 5 and 10 dimensions;
is the UMAP algorithm amenable to fit only once using `n_components=10` and then take
the first 2 or 5 components, or would this be complete nonsense and I should make separate fittings?

**Comment by adelejackson** [comment]
(Caveat: I'm not Leland McInnes; I'm a mathematician and am comfortable with the maths behind UMAP but am not familiar with the code. Take this with a grain of salt.)

I wouldn't expect this to work. Unlike in PCA, the coordinate system in the transformed data has no particular meaning -- you get the same cost for any isometry of a given representation (in particular, for a rotation). (There's also no reason the first two components would work any better than the last two.)

For example, say you have a 10 dimensional dataset. You can "transform" this under UMAP and you should get basically the same dataset back (up to an isometry, I think). Taking two components of this "transformation", in general we certainly do not get the embedding you would get with n_components=2 -- at least, I would hope not, or you could just restrict to two components instead of running UMAP!

**Comment by lmcinnes** [comment]
Thanks @adelejackson ! This one slipped between the cracks on me, so I greatly appreciate your expertise and assistance. Is there any chance I can talk you into further contributions and collaboration?

**Comment by adelejackson** [comment]
> I don't think, however, that you get the "same" dataset up to isometry when `n_components = n_features`; from what I understood from the paper the umap algo works with local distances, and it attempts to make a reconstruction that preserve the simplicial structure, i.e. it should preserve the topology of the dataset, not its metric.

Hmm, yes, now I think about it what I said isn't right. It should preserve the distances between points (up to scaling by some fixed factor) though, I think? (An embedding that does this would minimise the cost.) So something a lot weaker than an isometry of R^n, but that still respects distance between points.

**Comment by adelejackson** [comment]
> Thanks @adelejackson ! This one slipped between the cracks on me, so I greatly appreciate your expertise and assistance. Is there any chance I can talk you into further contributions and collaboration?

I'm definitely open to that :) I'm working in machine learning research before starting a maths PhD in low-dimensional topology in October, so this is all stuff I'm very interested in! I might send you an email about what I could do that would be most effective.

**Comment by adelejackson** [comment]
> I don't think, however, that you get the "same" dataset up to isometry when `n_components = n_features`; from what I understood from the paper the umap algo works with local distances, and it attempts to make a reconstruction that preserve the simplicial structure, i.e. it should preserve the topology of the dataset, not its metric.

Yep, you're completely correct; I forgot that we use the Euclidean metric for the low-dimensional representation, not the knn-weighted one.
